tests/monitoring/metrics.go

```diff
@@ -0,0 +1,150 @@
+
+		By("Deleting the VirtualMachine")
+		err = virtClient.VirtualMachine(vm.Namespace).Delete(context.Background(), vm.Name, &metav1.DeleteOptions{})
+		Expect(err).ToNot(HaveOccurred())
+
+		libmonitoring.WaitForMetricValue(virtClient, "kubevirt_number_of_vms", -1)
+	}
+
+	BeforeEach(func() {
+		virtClient = kubevirt.Client()
+
+		setupVM()
+
+		metrics, err = libmonitoring.QueryRange(virtClient, "{__name__=~\"kubevirt_.*\"}", time.Now().Add(-1*time.Minute), time.Now(), 15*time.Second)
+		Expect(err).ToNot(HaveOccurred())
+
+		Expect(metrics.Status).To(Equal("success"))
+		Expect(metrics.Data.ResultType).To(Equal("matrix"))
+		Expect(metrics.Data.Result).ToNot(BeEmpty(), "No metrics found")
+	})
+
+	containsMetric := func(metric operatormetrics.Metric, labels map[string]string) types.GomegaMatcher {
+		return &libmonitoring.MetricMatcher{Metric: metric, Labels: labels}
+	}
+
+	Context("Prometheus metrics", func() {
+		var excludedMetrics = map[string]bool{
+			// virt-api
+			// can later be added in pre-existing feature tests
+			"kubevirt_portforward_active_tunnels":  true,
+			"kubevirt_usbredir_active_connections": true,
+			"kubevirt_vnc_active_connections":      true,
+			"kubevirt_console_active_connections":  true,
+
+			// virt-controller
+			// needs a migration - ignoring since already tested in - VM Monitoring, VM migration metrics
+			"kubevirt_vmi_migration_phase_transition_time_from_creation_seconds": true,
+			"kubevirt_vmi_migrations_in_pending_phase":                           true,
+			"kubevirt_vmi_migrations_in_scheduling_phase":                        true,
+			"kubevirt_vmi_migrations_in_running_phase":                           true,
+			"kubevirt_vmi_migration_succeeded":                                   true,
+			"kubevirt_vmi_migration_failed":                                      true,
+		}
+
+		It("should contain virt-operator metrics", func() {
+			err = virtoperator.SetupMetrics()
+			Expect(err).ToNot(HaveOccurred())
+
+			for _, metric := range virtoperator.ListMetrics() {
+				Expect(metrics.Data.Result).To(ContainElement(containsMetric(metric, nil)))
+			}
+		})
+
+		It("should contain virt-api metrics", func() {
+			err = virtapi.SetupMetrics()
+			Expect(err).ToNot(HaveOccurred())
+
+			for _, metric := range virtapi.ListMetrics() {
+				if excludedMetrics[metric.GetOpts().Name] {
+					continue
+				}
+
+				Expect(metrics.Data.Result).To(ContainElement(containsMetric(metric, nil)))
+			}
+		})
+
+		It("should contain virt-controller metrics", func() {
+			err = virtcontroller.SetupMetrics(nil, nil, nil, nil, nil, nil, nil, nil)
+			Expect(err).ToNot(HaveOccurred())
+
+			for _, metric := range virtcontroller.ListMetrics() {
+				if excludedMetrics[metric.GetOpts().Name] {
+					continue
+				}
+
+				Expect(metrics.Data.Result).To(ContainElement(containsMetric(metric, nil)))
+			}
+		})
+
+		Context("perfscale histogram metrics", func() {
```

Comment:
Why can't we cover it with unit test instead?

Reply:
we would only test that they are registered in the component, not that they are correctly set up, and being exposed correctly
and since we already fetched the list of KubeVirt metrics, its just a simple validation